tests/quality_pragmaitc/test_refinement.py

```diff
@@ -24,63 +32,430 @@ def test_shouldBeApplicable():
-    assert True == delegation.isApplicable(fullContext)
+    assert True is goal.isApplicable(fullContext)
+    assert True is task.isApplicable(fullContext)
+    assert True is delegation.isApplicable(fullContext)
+
+
+def test_shouldBeNotApplicable():
+    goal = Goal()
+    task = Task()
+    delegation = Delegation()
+
+    context = Context("C1")
+
+    task.addApplicableContext(context)
+
+    goal.addApplicableContext(context)
+
+    delegation.addApplicableContext(context)
+
+    wrongContext = Context("C2")
+    fullContext = []
+    fullContext.append(wrongContext)
+
+    assert False is goal.isApplicable(fullContext)
+    assert False is task.isApplicable(fullContext)
+    assert False is delegation.isApplicable(fullContext)
+
+
+def test_aTaskShouldBeAchievable():
+    task = Task()
+
+    currentContext = Context("C1")
+    fullContext = []
+    fullContext.append(currentContext)
+
+    qc = QualityConstraint(
+        currentContext, CommonMetrics.SECONDS, 15, Comparison.LESS_OR_EQUAL_TO)
+
+    task.addApplicableContext(currentContext)
+    task.setProvidedQuality(currentContext, CommonMetrics.SECONDS, 12)
+
+    interp = Interpretation()
+    interp.addQualityConstraint(qc)
+
+    assert task in task.isAchievable(
+        fullContext, interp).getTasks()
+
+
+def test_shouldAddSeveralContextsAtOnce():
+    context1 = Context("C1")
+    context2 = Context("C2")
+
+    task = Task()
+    originalSize = 0
+
+    if None is task.getApplicableContext():
+        originalSize = len(task.getApplicableContext())
+
+    set = []
+
+    set.append(context1)
+    set.append(context2)
+
+    task.addApplicableContext(set)
+
+    assert 2 == len(task.getApplicableContext()) - originalSize
+
+
+def test_aNonApplicableRootGoalIsNotAchievable():
+    goal = Goal(False)
+    current = Context("C1")
+    fullContext = []
+
+    qc = QualityConstraint(current, CommonMetrics.SECONDS,
+                           15, Comparison.LESS_OR_EQUAL_TO)
+    goal.addApplicableContext(Context("C2"))
+
+    interp = Interpretation()
+    interp.addQualityConstraint(qc)
+
+    assert goal.isAchievable(fullContext, interp) is None
+
+
+def test_aGoalWithATaskMayBeAchievable():
+    goal = Goal(False)
+
+    task = Task()
+
+    current = Context("C1")
+    fullContext = []
+    fullContext.append(current)
+
+    qc = QualityConstraint(current, CommonMetrics.SECONDS,
+                           15, Comparison.LESS_OR_EQUAL_TO)
+    interp = Interpretation()
+    interp.addQualityConstraint(qc)
+
+    task.addApplicableContext(current)
+    task.setProvidedQuality(current, CommonMetrics.SECONDS, 13)
+
+    goal.addDependency(task)
+    goal.setIdentifier("Root")
+    goal.addApplicableContext(current)
+
+    plan = goal.isAchievable(fullContext, interp)
+    assert len(plan.getTasks()) == 1
+
+
+def test_aGoalAndDecomposedWithTwoTasksMayBeAchievable():
+    goal = Goal(Decomposition.AND)
+    assert not goal.isOrDecomposition
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Use of assert detected. The enclosed code will be removed when compiling to optimised byte code.](https://app.codacy.com/gh/lesunb/p2psimpy/pullRequest?prid=4861978)